### PR TITLE
Feature/gh 11154 remove value from input field

### LINKED ISSUE #11154: Not possible to remove value from the Input field
When user tries to remove the last character from the input field, it appears again after the update roundtrip.

### PR DESCRIPTION
**Pull request by AlexanderKuschner**
Allow removing of value from the input field for CPQ configurator;
Proper handling CPQ attributes of type READ_ONLY and INPUT NUMBER, which are currently not supported;
Use CPQ attribute name instead of label, if label is not provided.

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts [hunk 1]
```diff
@@ -108,6 +109,7 @@ const cpqAttribute2: Cpq.Attribute = {
   description: cpqAttributeDescription2,
   label: cpqAttributeLabel2,
   displayAs: cpqAttributeDisplayAs2,
+  dataType: cpqAttributeDataType2,
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts]: In my opinion it does not make sense to define a constant if it is used exactly once. Also the name 'cpqAttributeDataType2' is not very descriptive. I would prefer to inline this constant, as well as all other constants in the file that are used only once.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts [hunk 2]
```diff
@@ -605,6 +626,20 @@ describe('CpqConfiguratorNormalizer', () => {
...
+        stdAttrCode: 2,
+        displayAs: Cpq.DisplayAs.INPUT,
+        dataType: Cpq.DataType.INPUT_NUMBER,
+        isEnabled: true,
+        values: [],
+      };
+      expect(cpqConfiguratorNormalizer.convertAttributeType(cpqAttr)).toBe(
+        Configurator.UiType.NOT_IMPLEMENTED
+      );
+    });
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts:622]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts [hunk 3]
```diff
@@ -460,6 +462,24 @@ describe('CpqConfiguratorNormalizer', () => {
...
+    cpqConfiguratorNormalizer.convertAttribute(
+      cpqAttributeWithoutLabel,
+      cpqGroupId,
+      CURRENCY,
+      attributeList
+    );
+    const attribute: Configurator.Attribute = attributeList[0];
+    expect(attributeList.length).toBe(1);
+    expect(attribute.label).toBe('AttributeName');
+  });
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts:462]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts [hunk 4]
```diff
@@ -126,7 +126,9 @@ export class CpqConfiguratorNormalizer
       attrCode: sourceAttribute.stdAttrCode,
       name: sourceAttribute.pA_ID.toString(),
       description: sourceAttribute.description,
-      label: sourceAttribute.label,
+      label: sourceAttribute.label
+        ? sourceAttribute.label
+        : sourceAttribute.name,
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts:131]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts [hunk 5]
```diff
@@ -303,7 +305,11 @@ export class CpqConfiguratorNormalizer
       }
 
       case Cpq.DisplayAs.INPUT: {
-        uiType = Configurator.UiType.STRING;
+        if (sourceAttribute?.dataType === Cpq.DataType.INPUT_STRING) {
+          uiType = Configurator.UiType.STRING;
+        } else {
+          uiType = Configurator.UiType.NOT_IMPLEMENTED;
+        }
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts:312]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.spec.ts [hunk 6]
```diff
@@ -204,12 +204,23 @@ describe('CpqConfiguratorOverviewNormalizer', () => {
...
+  it('should map user input with not supported DataType as NOT_IMPLEMENTED', () => {
+    attr.userInput = 'input';
+    attr.displayAs = Cpq.DisplayAs.INPUT;
+    attr.dataType = Cpq.DataType.INPUT_NUMBER;
+    const ovAttrs = serviceUnderTest['convertAttribute'](attr, CURRENCY);
+    expect(ovAttrs.length).toBe(1);
+    expect(ovAttrs[0].value).toEqual('NOT_IMPLEMENTED');
+    expect(ovAttrs[0].productCode).toBeUndefined();
+  });
+
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.spec.ts:223]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.ts [hunk 7]
```diff
@@ -78,12 +78,15 @@ export class CpqConfiguratorOverviewNormalizer
...
     switch (attr.displayAs) {
       case Cpq.DisplayAs.INPUT:
-        if (attr.userInput && attr.userInput.length > 0) {
-          ovValues.push(this.extractOvValueUserInput(attr, currency));
+        if (attr?.dataType === Cpq.DataType.INPUT_STRING) {
+          if (attr.userInput && attr.userInput.length > 0) {
+            ovValues.push(this.extractOvValueUserInput(attr, currency));
+          }
+        } else {
+          ovValues.push({ attribute: undefined, value: 'NOT_IMPLEMENTED' });
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.ts:86]: this does not seem to be related to 'move value from input field'. did you fix multiple issues with this PR?

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
there a lot of additional changes, besides fixing the reported bug. was this intended?

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts (hunk 2 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts:622]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts (hunk 3 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts:462]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts (hunk 4 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts:131]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts (hunk 5 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.ts:312]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.spec.ts (hunk 6 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.spec.ts:223]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.ts (hunk 7 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-overview-normalizer.ts:86]: Yes. Several issues (directly or indirectly related to handling of input fields) are fixed with this pull request. All adjustments are mentioned in the pull request comment.

File: feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts (hunk 1 above)
Comment (AlexanderKuschner) [reply feature-libs/product-configurator/rulebased/cpq/cpq-configurator-normalizer.spec.ts]: The const name cpqAttributeDataType2 means that it is used in the attribute2 rather than a 'special' data type. I agree, that it is better to not introduce this const and to inline the related value in the attribute definition directly.  Such constants are removed now.

**Review by Uli-Tiger — APPROVED** [review]
